test/unit/kubernetes-deploy/kubernetes_resource/deployment_test.rb

```diff
@@ -2,118 +2,332 @@
+      replica_sets: replica_sets
+    )
+    refute deploy.deploy_succeeded?
   end
 
-  def test_deploy_succeeded_fails_with_max_unavailable
-    rollout = {
-      'metadata' => {
-        'name' => 'fake',
-        'annotations' => { KubernetesDeploy::Deployment::REQUIRED_ROLLOUT_ANNOTATION => 'maxUnavailable' }
-      }
+  def test_deploy_succeeded_with_max_unavailable_as_percent
+    deployment_status = {
+      "replicas" => 3,
+      "updatedReplicas" => 2,
+      "unavailableReplicas" => 2,
+      "availableReplicas" => 1
+    }
+
+    rs_status = {
+      "replicas" => 2,
+      "availableReplicas" => 1,
+      "readyReplicas" => 1
     }
+    replica_sets = [build_rs_template(status: rs_status)]
+
+    dep_template = build_deployment_template(status: deployment_status,
+      rollout: 'maxUnavailable', max_unavailable: '100%')
+    deploy = build_synced_deployment(template: dep_template, replica_sets: replica_sets)
+    assert deploy.deploy_succeeded?
+
+    # rounds up to two max
+    deploy = build_synced_deployment(
+      template: build_deployment_template(status: deployment_status, rollout: 'maxUnavailable', max_unavailable: '67%'),
+      replica_sets: replica_sets
+    )
+    assert deploy.deploy_succeeded?
 
-    deploy = KubernetesDeploy::Deployment.new(namespace: "", context: "", logger: logger, definition: rollout)
-    mock_rs = Minitest::Mock.new
-    needed = 2
-    mock_rs.expect :present?, true
-    mock_rs.expect :desired_replicas, needed
-    mock_rs.expect :ready_replicas, needed - 1
-    mock_rs.expect :available_replicas, needed - 1
-    deploy.instance_variable_set(:@max_unavailable, 0)
-    deploy.instance_variable_set(:@latest_rs, mock_rs)
-    deploy.instance_variable_set(:@desired_replicas, needed)
+    # rounds down to one max
+    deploy = build_synced_deployment(
+      template: build_deployment_template(status: deployment_status, rollout: 'maxUnavailable', max_unavailable: '66%'),
+      replica_sets: replica_sets
+    )
+    refute deploy.deploy_succeeded?
 
+    deploy = build_synced_deployment(
+      template: build_deployment_template(status: deployment_status, rollout: 'maxUnavailable', max_unavailable: '0%'),
+      replica_sets: replica_sets
+    )
     refute deploy.deploy_succeeded?
   end
 
-  def test_deploy_succeeded_fails_with_max_unavailable_as_a_percent
-    rollout = {
-      'metadata' => {
-        'name' => 'fake',
-        'annotations' => { KubernetesDeploy::Deployment::REQUIRED_ROLLOUT_ANNOTATION => 'maxUnavailable' }
-      }
-    }
+  def test_deploy_succeeded_raises_with_invalid_rollout_annotation
+    deploy = build_synced_deployment(
+      template: build_deployment_template(rollout: 'bad'),
+      replica_sets: [build_rs_template]
+    )
+    msg = "'#{KubernetesDeploy::Deployment::REQUIRED_ROLLOUT_ANNOTATION}: bad' is "\
+      "invalid. Acceptable values: maxUnavailable, full, none"
```

Comment:
Can you use `REQUIRED_ROLLOUT_TYPES` here instead of `maxUnavailable, full, none`.

Reply:
I can if you feel strongly, but my opinion is that the further these get from showing the human-readable message, the less valuable the assertion is. The fact that they were written out made me notice that our `join` was missing a space.